vms/platformvm/state/state.go

```diff
@@ -2727,82 +2722,82 @@ func writePendingDiff(
 }
 
 func (s *state) writeSubnetOnlyValidators() error {
-	// Write modified weights:
+	// Write modified weights
 	for subnetID, weight := range s.sovDiff.modifiedTotalWeight {
-		if err := database.PutUInt64(s.weightsDB, subnetID[:], weight); err != nil {
-			return err
-		}
-	}
-	maps.Clear(s.sovDiff.modifiedTotalWeight)
-
-	sovChanges := s.sovDiff.modified
-	// Perform deletions:
-	for validationID, sov := range sovChanges {
-		if !sov.isDeleted() {
-			// Additions and modifications are handled in the next loops.
-			continue
-		}
-
-		// The next loops shouldn't consider this change.
-		delete(sovChanges, validationID)
-
-		subnetIDNodeID := subnetIDNodeID{
-			subnetID: sov.SubnetID,
-			nodeID:   sov.NodeID,
+		var err error
+		if weight == 0 {
+			err = s.weightsDB.Delete(subnetID[:])
+		} else {
+			err = database.PutUInt64(s.weightsDB, subnetID[:], weight)
 		}
-		subnetIDNodeIDKey := subnetIDNodeID.Marshal()
-		err := s.subnetIDNodeIDDB.Delete(subnetIDNodeIDKey)
 		if err != nil {
 			return err
 		}
 
-		priorSOV, wasActive := s.activeSOVLookup[validationID]
-		if wasActive {
-			delete(s.activeSOVLookup, validationID)
-			s.activeSOVs.Delete(priorSOV)
+		s.weightsCache.Put(subnetID, weight)
+	}
+
+	// The SoV diff application is split into two loops to ensure that all
+	// deletions to the subnetIDNodeIDDB happen prior to any additions.
+	// Otherwise replacing an SoV by deleting it and then re-adding it with a
+	// different validationID could result in an inconsistent state.
+	for validationID, sov := range s.sovDiff.modified {
+		// Delete the prior validator if it exists
+		var err error
+		if s.activeSOVs.delete(validationID) {
```

Comment:
can we use get first and then delete  from activeSovs if it's indeed active or delete from inactive? I think `delete` does a bit too much work here and also wrapping `get`.

Reply:
I actually feel like it's pretty normal for `delete` to return whether or not it deleted something. For example, this is the behavior of `tree.Delete`

Reply:
I just found it a bit weird that we try to immediately delete it from activeSOVs before we check if it's there. But I think it's mostly a preference so should be fine.